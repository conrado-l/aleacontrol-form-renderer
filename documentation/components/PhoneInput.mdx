---
name: PhoneInput
---

import { Playground, PropsTable } from 'docz'
import PhoneInput from '../../src/components/PhoneInput.js'

# PhoneInput

<PropsTable of={PhoneInput} />

## Description
A select input for the extension and an string input for the phone
## Demo and usage

<Playground>
    <PhoneInput value={{phone: '123456', ext:'387'}} name='phone' options={[{value: 387, description: 'Bosnia'}, {value: 54, description: 'Argentina'}]} col={4} update={(e) => e}/>
</Playground>
