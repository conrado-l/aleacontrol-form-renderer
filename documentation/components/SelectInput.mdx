---
name: SelectInput
---

import { Playground, PropsTable } from 'docz'
import SelectInput from '../../src/components/SelectInput.js'

# SelectInput

<PropsTable of={SelectInput} />

## Description
A radio input with options
## Demo and usage

<Playground>
    <SelectInput value='male' name='gender' options={[{value: 'male', description: 'Male'}, {value: 'female', description: 'Female'}]} col={4} update={(e) => e}/>
</Playground>
