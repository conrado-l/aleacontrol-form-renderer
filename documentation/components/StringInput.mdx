---
name: StringInput
---

import { Playground, PropsTable } from 'docz'
import StringInput from '../../src/components/StringInput.js'

# StringInput

<PropsTable of={StringInput} />

## Description
A simple text input
## Demo and usage

<Playground>
    <StringInput value='Hello' name='greeting' col={4} update={(e) => e}/>
</Playground>
