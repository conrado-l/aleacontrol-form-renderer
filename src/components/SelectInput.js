import React, { Component } from 'react'
import PropTypes from 'prop-types'

class SelectInput extends Component {
  render () {
    // Generate the select options
    const options = this.props.options.map((input) => {
      return <option value={input.value} key={input.value}> {input.description}</option>
    })

    return (
      <select className={`form-control col-xs-12 col-${this.props.col}`} onChange={(e) => this.props.update(this.props.name, e.target.value)}>
        {options}
      </select>
    )
  }
}

SelectInput.propTypes = {
  options: PropTypes.array,
  name: PropTypes.string,
  col: PropTypes.number,
  update: PropTypes.func
}
export default SelectInput
