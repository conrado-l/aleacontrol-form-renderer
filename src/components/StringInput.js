import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Form from 'react-bootstrap/Form'

/**
 * Regular text type input
 */
class StringInput extends Component {
  constructor (props) {
    super(props)
    this.state = {
      value: props.value
    }

    this.handleChange = this.handleChange.bind(this)
  }

  /**
   * Updates the input value and sends the value through the parent callback function
   * @param {object} e Event
   */
  handleChange (e) {
    const newValue = e.target.value
    this.setState({ value: newValue })
    this.props.update(this.props.name, newValue)
  }

  render () {
    return (
      <Row>
        <Col xs={12} sm={this.props.col}>
          <Form.Control
            type='text'
            name={this.props.name}
            value={this.state.value || ''}
            onChange={this.handleChange} />
        </Col>
      </Row>
    )
  }
}

StringInput.propTypes = {
  value: PropTypes.string,
  name: PropTypes.string,
  col: PropTypes.number,
  update: PropTypes.func
}

export default StringInput
