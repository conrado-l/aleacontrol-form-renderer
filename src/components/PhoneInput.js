import React, { Component } from 'react'
import PropTypes from 'prop-types'
import SelectInput from './SelectInput'
import StringInput from './StringInput'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'

class PhoneInput extends Component {
  constructor (props) {
    super(props)

    this.state = {
      ext: props.value.ext,
      phone: props.value.phone
    }

    this.updatePhone = this.updatePhone.bind(this)
  }

  /**
   * Updates state, wait for the state update and send the new phone object to the parent
   * @param inputKey
   * @param inputValue
   */
  updatePhone (inputKey, inputValue) {
    this.setState({ [inputKey]: inputValue }, () => {
      this.props.update('phone', this.state)
    })
  }

  render () {
    return (
      <Row>
        <Col xs={12} sm={this.props.col}>
          <SelectInput name='ext' value={this.state.ext} options={this.props.options} update={this.updatePhone} />
        </Col>
        <Col xs={12} sm={this.props.col}>
          <StringInput name='phone' value={this.state.phone} update={this.updatePhone} />
        </Col>
      </Row>
    )
  }
}

PhoneInput.propTypes = {
  value: PropTypes.object,
  options: PropTypes.array,
  col: PropTypes.number,
  update: PropTypes.func
}

export default PhoneInput
