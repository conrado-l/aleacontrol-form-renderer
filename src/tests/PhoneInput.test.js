import React from 'react'
import { shallow } from 'enzyme'
import PhoneInput from '../components/PhoneInput'
import SelectInput from '../components/SelectInput'
import StringInput from '../components/StringInput'
import renderer from 'react-test-renderer'

describe('PhoneInput', () => {
  const extOptions = [
    { value: '387', description: 'Bosnia and Herzegovina' },
    { value: '54', description: 'Argentina' }
  ]

  it('should render correctly phone component', () => {
    const SelectInputComponent = renderer
      .create(<SelectInput value={{ phone: '123456', ext: '387' }} options={extOptions} />)
      .toJSON()
    expect(SelectInputComponent).toMatchSnapshot()
  })

  it('should render the ext select', () => {
    const PhoneInputComponent = shallow(<PhoneInput value={{ phone: '123456', ext: '387' }} options={extOptions} />)
    expect(PhoneInputComponent.find(SelectInput)).toHaveLength(1)
  })

  it('should render the phone input', () => {
    const PhoneInputComponent = shallow(<PhoneInput value={{ phone: '123456', ext: '387' }} options={extOptions} />)
    expect(PhoneInputComponent.find(StringInput)).toHaveLength(1)
  })

  it('should update the state when the updatePhone callback is executed', () => {
    const updateFn = jest.fn()
    const PhoneInputComponent = shallow(<PhoneInput value={{ phone: '123456', ext: '387' }} options={extOptions} update={updateFn} />)
    const PhoneInputInstance = PhoneInputComponent.instance()

    PhoneInputInstance.updatePhone('ext', '54')
    PhoneInputInstance.updatePhone('phone', '654321')

    expect(PhoneInputComponent.state().ext).toBe('54')
    expect(PhoneInputComponent.state().phone).toBe('654321')
  })

  it('should emit the local state to the parent through the update callback when the phone/ext is updated', () => {
    const updateFn = jest.fn()
    const PhoneInputComponent = shallow(<PhoneInput value={{ phone: '123456', ext: '387' }} options={extOptions}
      update={updateFn} />)
    const PhoneInputInstance = PhoneInputComponent.instance()

    PhoneInputInstance.updatePhone('ext', '54')
    PhoneInputInstance.updatePhone('phone', '654321')

    expect(updateFn).toHaveBeenCalledTimes(2)

    expect(updateFn.mock.calls[0][0]).toBe('phone')
    expect(updateFn.mock.calls[0][1]).toEqual({ ext: '54', phone: '123456' })
    expect(updateFn.mock.calls[1][0]).toBe('phone')
    expect(updateFn.mock.calls[1][1]).toEqual({ ext: '54', phone: '654321' })
  })
})
