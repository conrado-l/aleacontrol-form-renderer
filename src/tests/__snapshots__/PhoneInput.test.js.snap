// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneInput should render correctly phone component 1`] = `
<select
  className="form-control col-xs-12 col-undefined"
  onChange={[Function]}
>
  <option
    value="387"
  >
     
    Bosnia and Herzegovina
  </option>
  <option
    value="54"
  >
     
    Argentina
  </option>
</select>
`;
